lib/adb-device.js

```diff
@@ -17,16 +41,366 @@ class ADBDevice {
+    // open a SYNC stream on the device for pushing the file
+    let packet = await this.sendAndOkay(ADB_COMMANDS.CMD_OPEN
+                                       , 12345
+                                       , 0
+                                       , syncBuf);
+    // this is our local id, the devices remote id
+    let localId = packet.arg2;
+    // this is our remote id, the devices local id
+    let remoteId = packet.arg1;
+    await this.stat(destination, localId, remoteId);
+    let devicePath = `${destination}/${getFileName(source)}`;
+    let sendBuffer = new Buffer(8);
+    sendBuffer.writeUInt32LE(ADB_SUBCOMMANDS.CMD_SEND, 0);
+    sendBuffer.writeUInt32LE(devicePath.length + 6, 4);
+    packet = await this.sendAndOkay(ADB_COMMANDS.CMD_WRTE
+                                   , localId
+                                   , remoteId
+                                   , sendBuffer);
+    await this.sendFile(source, destination, localId, remoteId);
+  }
+
+  // runs ADB pull to pull a file from the device to the local machine
+  async pull (source, destination) {
+    let syncBuf = new Buffer("sync:.");
+    // open a SYNC stream on the device for pulling a file from the device
+    let packet = await this.sendAndOkay(ADB_COMMANDS.CMD_OPEN
+                                       , 12345
+                                       , 0
+                                       , syncBuf);
+    // this is our local id, the devices remote id
+    let localId = packet.arg2;
+    // this is our remote id, the devnices local id
+    let remoteId = packet.arg1;
+    // tell the device we're going to want stats about a file
+    packet = await this.stat(source, localId, remoteId);
+    if (packet.data.readUInt32LE(4) === 0) { // file doesn't exist
+      console.log("The file you're trying to pull doesn't exist.");
+      await this.close(localId, remoteId);
+      return;
+    }
+
+    // tell the device we want to recv the file
+    let recvBuffer = new Buffer(8);
+    recvBuffer.writeUInt32LE(ADB_SUBCOMMANDS.CMD_RECV, 0);
+    recvBuffer.writeUInt32LE(source.length, 4);
+    packet = await this.sendAndOkay(ADB_COMMANDS.CMD_WRTE
+                                   , localId
+                                   , remoteId
+                                   , recvBuffer);
+    // send the path to the file want to pull again
+    packet = await this.sendAndOkay(ADB_COMMANDS.CMD_WRTE
+                                   , localId
+                                   , remoteId
+                                   , source);
+    console.log("sent final source");
+    // recv the file data from the device
+    let rawData = await this.recvFile(localId, remoteId);
+    // pull the file data out (there's data related to the protocol in rawData)
+    let fileData = parseFileData(rawData);
+    // write data to a local file
+    let writeAsync = Promise.promisify(fs.writeFile);
+    await writeAsync(destination, fileData); // do we need to await this line?
```

Comment:
yes that is right. You would need to keep track of threads that you spawn off. You can try using some thing like parallel, never tried it but [await*](https://github.com/jaydson/es7-async#the-new-awesome-beast-async-functions) looks interesting.